R/import_ebird.R

```diff
@@ -138,9 +152,10 @@ arrow_open_ebird_txt <- function(ebd, dest) {
 
 ebird_col_type <- function(col_names) {
   # types for columns that are not character
-  col_types <- c(`LAST EDITED DATE` = "timestamp", 
-                 `TAXONOMIC ORDER` = "integer", 
-                 `LATITUDE` = "double", `LONGITUDE` = "double", 
+  col_types <- c(`LAST EDITED DATE` = "string", ## Obviously not, but fails on relJuly-2021
```

Comment:
I see you've made this change to the data types, but aren't calling `ebird_col_type()` anywhere anymore, so the change doesn't have an impact. I think arrow is mostly guessing the column types correctly, but I worry about some that have a large number of missing values, e.g. in the example dataset these columns are entirely empty, so it's putting the type as `null`:
```
county: null
county_code: null
iba_code: null
bcr_code: null
usfws_code: null
atlas_block: null
```

Reply:
yup, we should be enforcing column type; guessing we should be able to do that directly from the first `arrow::open_dataset()`, and can drop the deprecated `arrow_open_ebird_txt()`. Sound good?

Reply:
yeah, I think that makes sense

Reply:
hmm, now I recall why I didn't do this before...

Minor bit, we need to detect if we have checklist or observation data so that we use the right schema.

More technical problem, I'm not having any luck reading in the schema (see 94e07771f9957197c9b952004b4b503f17b6394d) -- arrow complains about any column that I have not typed as `arrow::string()`.... Not sure what is up. e.g. on the branch `patch-schema` I get:


```
> birddb::import_ebird(birddb::sample_observation_data())
eBird observations data already exists in BIRDDB_HOME, would you like to overwrite this data? (yes/no/Cancel) yes
Importing observations data from the eBird Basic Dataset: ebd_relAug-2021.tar

 Error: Invalid: Could not open CSV input source '/tmp/RtmpSwrma7/ebird_tmp2ae3c332f54d20/ebd_HK_201201_201212_relAug-2021.txt.gz': Invalid: In CSV column #1: Row #1: CSV conversion error to date64[ms]: invalid value 'LAST EDITED DATE'
``` 
making the date a string gets it passed this column, only to fail on the next non-string column

Reply:
That's strange, a bug in arrow maybe? I'm getting the same issue with the CRAN version.

I was able to fix the blank column error I was getting in the other PR (`"Invalid: CSV parse error: Row #1: Expected 47 columns, got 48"`), albeit in a hacky way. https://github.com/cboettig/birddb/pull/9/commits/82316b83bbf569452d2c30fb7143f1e02b798d06

But now I'm stuck on the same issue you're having here arrow requiring everything be a string.

Reply:
Thanks for confirming.  For the unnamed row, using an unnamed element in the schema seemed to work for me:

https://github.com/cboettig/birddb/blob/patch-schema/R/schemas.R#L51

let's see if we can strip this back to a more minimal example and then we can file a bug report.  

I wish there was more control over the column type guessing behavior as well.  In `readr`, for instance, it's possible to either alter the default guess used when a column is all-missing-data (`col_types = readr::cols(.default = "c")`), and you also have the option to increase the number of lines used to formulate the guess.  (further, readr samples 1000 lines at random now, rather than using the first 1000, to make a guess; which apparently helps a lot too).